doc/specs/#888 - Com Api.md

```diff
@@ -0,0 +1,851 @@
+And there is Windows.Management.Deployment.PackageManager which allows packages with the packageManagement
+capability to install msix apps and uses "Find" to describe looking up an app
+https://docs.microsoft.com/en-us/uwp/api/windows.management.deployment.packagemanager?view=winrt-19041
+
+I've chosen to align with the Windows APIs in using \*Catalog and Find. But this may cause some confusion if
+callers are using both the winget command line and this API. In particular a problem may be that since this
+API does not yet propose to implement adding a "source", client applications would need to work with the
+command line interface in order to do that which may cause the name change from Source to AppCatalog to be
+particularly confusing.
+
+# 5 API Details
+
+```c# (but really MIDL3)
+namespace Microsoft.Management.Deployment
+{
+    [contractversion(1)]
+    apicontract WindowsPackageManagerContract{};
+
+    /// State of the install.
+    [contract(Microsoft.Management.Deployment.WindowsPackageManagerContract, 1)]
+    enum AppInstallProgressState
+    {
+        /// The install is queued but not yet active. Cancellation of the IAsyncOperationWithProgress in this
+        /// state will prevent the app from downloading or installing.
+        Queued,
+        /// The installer is downloading. Cancellation of the IAsyncOperationWithProgress in this state will
+        /// end the download and prevent the app from installing.
+        Downloading,
+        /// The install is in progress. Cancellation of the IAsyncOperationWithProgress in this state will not
+        /// stop the installation or the post install cleanup.
+        Installing,
+        /// The installer has completed and cleanup actions are in progress. Cancellation of the
+        /// IAsyncOperationWithProgress in this state will not stop cleanup or roll back the install.
+        PostInstall,
+        /// The operation has completed.
+        Finished,
+    };
+
+    /// Progress object for the install
+    /// DESIGN NOTE: percentage for the install as a whole is purposefully not included as there is no way to
+    /// estimate progress when the installer is running.
+    [contract(Microsoft.Management.Deployment.WindowsPackageManagerContract, 1)]
+    struct InstallProgress
+    {
+        /// State of the install
+        AppInstallProgressState State;
+        /// DESIGN NOTE: BytesDownloaded may only be available for downloads done by Windows Package Manager itself.
+        /// Number of bytes downloaded if known
+        UInt64 BytesDownloaded;
+        /// DESIGN NOTE: BytesRequired may only be available for downloads done by Windows Package Manager itself.
+        /// Number of bytes required if known
+        UInt64 BytesRequired;
+        /// Download percentage completed
+        Single DownloadPercentage;
+    };
+
+    /// Result of the install
+    [contract(Microsoft.Management.Deployment.WindowsPackageManagerContract, 1)]
+    runtimeclass InstallResult

```

Comment:
Consider adding the install's final ErrorCode here.

Reply:
Also, a few months ago there was some discussion about "unrecoverable" errors, for which we should not allow the user to retry the installation. Are we still planning to surface the "Do Not Retry" result? (Example: OWC Orchestrator asks winget to download a product, and winget finds that the product is on a disallowed list. We can make sure the user doesn't repeatedly click the Install button and show a better error message if we have the "Do Not Retry" flag.)

Reply:
@kennykerr @jonwis what's the normal way for apis to return something like this? I do think it would be nice from an error handling perspective to split the error codes out the way msix currently does it with a generic "install failed" error that everyone can handle, and then an ExtendedError property on the result with the exact issue that was hit. But it looked like if IAsyncOperation.AsyncStatus is AsyncStatus::Error that the Result object isn't expected to be valid which makes sense to me. Is the preference to let it go to AsyncStatus::Complete (as in, the attempt completed) and then add error code properties that can be checked to see if it succeeded. Or to use the provisional result feature where you can set the result with the error codes and then put it into the Error state?

Reply:
It seems suboptimal that the caller needs to catch two different exception types when calling InstallPackage (cancel & hresult), then also check the async status and the install result.

Reply:
Most languages including C++, C#, and Rust will simply propagate the error information when you wait for the result of the async operation, either synchronously or asynchronously. No need to manually check the async status. In C++ you can simply `co_await` the async operation to cooperatively wait, or call `async.get()` to perform a blocking wait. On successful completion the result will be returned. On failure, the error will be rethrown.

Reply:
RECOMMEND: Follow -Result and ExtendedError patterns here